enhancements/workload-partitioning/wide-availability-workload-partitioning.md

```diff
@@ -0,0 +1,731 @@
+the same, we simply will allow you to apply partitioning on non single node
+topologies.
+
+We will use the global identifier to correctly modify the pod spec with the
+`requests.cpu` for the new `requests[management.workload.openshift.io/cores]`
+that are used by the workload partitioning feature.
+
+However, for Single-Node we will continue to check the conventional way to be
+able to support the upgrade flow from 4.11 -> 4.12. After 4.12 release that
+logic should no longer be needed and will be removed.
+
+### Node Admission Plugin
+
+We want to ensure that nodes that are not setup for CPU Partitioning do not get
+added to a cluster that is designated for CPU Partitioning. We will create an
+admission plugin for Nodes that will validate if a cluster is setup for CPU
+Partitioning then the node MUST contain a Capacity resource for
+`workload.openshift.io/cores`. All node creation requests from Kubelet currently add
+that information on boot up when registering with the API Server, we will
+leverage that to ensure a CPU Partitioned cluster only contains nodes created
+for CPU Partitioning.
+
+We will also keep in mind upgrades from Single Node clusters which already
+contain this feature. During initial upgrade, Single Node clusters will not
+contain the CPUPartitioningMode, for this reason we will fall back to checking
+with the old logic for Single Node to insure we do not cause issues when
+upgrading. This check for Single Node should be something that happens on initial
+upgrades, as NTO will update the `Infra.Status.CPUPartitioningMode` to the correct
+value after initial boot.
+
+### Performance Profile Controller
+
+Currently workload partitioning involves configuring CRI-O and Kubelet earlier
+in the processes as a separate machine config manifest that requires the same
+information present in the `PerformanceProfile` resource, that being the
+`isolated` and `reserved` CPU sets. Because configuring multiple resources with
+the right CPU sets consistently is error prone, we want to extend the
+PerformanceProfile API to include settings for workload partitioning.
+
+When installation is done and workload partitioning is set then from that point
+on the `kubelet` and `crio` only need to be configured with the desired CPU set
+to use. We currently express this to customers via the `reserved` CPU set as
+part of the performance profile api.
+
+We want to add a new `workloads` field to the `cpu` field that contains a list
+of enums for defining which workloads to pin. This should allow us to expand
+this in the future if desired, but in this enhancement we will only support
+`Infrastructure` which defines all of the Openshift workloads.
+
+```yaml
+apiVersion: performance.openshift.io/v2
+kind: PerformanceProfile
+metadata:
+  name: openshift-node-workload-partitioning-custom
+spec:
+  cpu:
+    isolated: 2-3
+    reserved: 0,1
+    # New addition
+    workloads:
+      - Infrastructure
```

Comment:
I like that idea, I'm not married to the `Infrastructure` enum, originally I had that level of implementation, but was not sure if `Management` was clear wording for an API to users, thus the grouping of `Infrastructure` as one way to encompass our workloads. The current way I have it implemented makes it trivial to support either path. @dhellmann do you have any thoughts on using `Management` instead of `Infrastructure`?

Reply:
I agree it makes sense to link the names here directly to the names in pod labels.

Is there actually a need to give the user control over different workloads beyond the management workload?

Reply:
Sounds good, I'll make the needed changes

Reply:
I wonder if we need to do this part of the work at all right now? Does the user need that much control right now?

Reply:
There must be at least a boolean and we decided that booleans should not be used. This can't be the default of course.

Reply:
I would say we don't allow custom enums here just yet, we just change the enum to `Management` to align with the annotation naming for clarity.